README.md
# lol_colors_flutter

Use lol Colors as a normal Color instead of using *Colors.blue* for example, you'll use **LolColors.c4714_1**

###### The convection used is:

LolColors.c + [**color number**] +  [**color position**]

In which **color number** is the color in the `lol Color List`, and the **color position** correspond to 1,2,3 or 4

For example

```dart
Text("Lol Colors",
  style: TextStyle(
    color : LolColors.c2877_3
  )
),
Container(color:LolColors.c2877_1),
Container(color:LolColors.c2877_2),
Container(color:LolColors.c2877_3),
Container(color:LolColors.c2877_4),
```

Also, you can used in Gradient, by accesing LolColors.colors_**color_number**, as is showed belown

```dart
Container(
  decoration : BoxDecoration(
    gradient: LinearGradient(
      colors : LolColors.colors_c2877
    )
  )
)
```

## Simple Demo

![demo v1](https://raw.githubusercontent.com/SergioEric/lol_colors_flutter/master/resources/gifs/demo-v1.gif)


## Lol Color List

.
![lol_colors_list_1](https://raw.githubusercontent.com/SergioEric/lol_colors_flutter/master/resources/list-lol-colors/lol_colors_list_1.png)

![lol_colors_list_1](https://raw.githubusercontent.com/SergioEric/lol_colors_flutter/master/resources/list-lol-colors/lol_colors_list_2.png)





project/CHANGELOG.md
## [0.0.1] - First 5 Colors

* Colors 4714, 2209, 2877, 2046, 1477

  
## [0.1.0] - Color Extension

* isDark()
* contrast()
* level()
* negate()


project/test/color_extension_test.dart
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:lol_colors_flutter/color_extension.dart';
void main() {
  // testWidgets('color extension ...', (tester) async {
  //   // TODO: Implement test
  // });
  group("ColorExtension", (){
    test("Test AA level on ColorExtension", (){
      // ARRANGE
      final Color background =  Colors.blue; // light color
      final Color foreground = Colors.black; // dark color
      // ACT
      String level = background.level(foreground);
      //ASSERT

      expect(level, "AA");
    });
    test("Test AAA level on ColorExtension", (){
      // ARRANGE
      final Color background =  Color(0xff21a0f3); // light color
      final Color foreground = Colors.black; // dark color
      // ACT
      String level = background.level(foreground);
      //ASSERT

      expect(level, "AAA");
    });
    test("Test Fail level on ColorExtension", (){
      // ARRANGE
      final Color background =  Color(0xff212cf3); // light color
      final Color foreground = Colors.black; // dark color
      // ACT
      String level = background.level(foreground);
      //ASSERT

      expect(level, "");
    });

  });
}


project/test/lol_colors_flutter_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:lol_colors_flutter/lol_colors_flutter.dart';

void main() {
  // test('adds one to input values', () {
  //   final calculator = Calculator();
  //   expect(calculator.addOne(2), 3);
  //   expect(calculator.addOne(-7), -6);
  //   expect(calculator.addOne(0), 1);
  //   expect(() => calculator.addOne(null), throwsNoSuchMethodError);
  // });
}


project/lib/color_extension.dart
import 'package:flutter/material.dart';
// import 'dart:math' as math;

/// TODO: add description
extension ColorExtension on Color {
  // YIQ equation from http://24ways.org/2010/calculating-color-contrast
  bool isDark() {
    final Color color = this;

    final int r = color.red;
    final int g = color.green;
    final int b = color.blue;

    final yiq = ((r * 299) + (g * 587) + (b * 114)) / 1000;
    //(yiq >= 128) ? 'black' : 'white';
    return (yiq >= 128);
  }

  bool isLight() {
    return !this.isDark();
  }

  double contrast(Color color2) {
    assert(color2 != null,
        "You must provide a second Color, Colors.n.contrast(color2)");
    // http://www.w3.org/TR/WCAG20/#contrast-ratiodef
    final double lum1 = this.computeLuminance();
    final double lum2 = color2.computeLuminance();

    if (lum1 > lum2) {
      return (lum1 + 0.05) / (lum2 + 0.05);
    }

    return (lum2 + 0.05) / (lum1 + 0.05);
  }

  String level(Color color2) {
    assert(color2 != null, "You must provide a second Color");

    final double contrastRatio = this.contrast(color2);
    if (contrastRatio >= 7.1) {
      return 'AAA';
    }
    // '' means contrast fail
    return (contrastRatio >= 4.5) ? 'AA' : '';
  }

  Color negate([int opacity = 255]) => Color.fromARGB(
      opacity, 255 - this.red, 255 - this.green, 255 - this.blue);

  String toRBG() {
    return "rgb(${this.red},${this.green},${this.blue})";
  }

  //TODO create a function that return the color primary value shade
  //TODO create a function that evaluate the best level of a color between white and black
}


project/example/lib/drop.widget.dart
import 'package:flutter/material.dart';

/// DropColor widget draw a drop with the color pass, 
/// accept a **DropRatio** that can be optional
/// 
/// Size by default is DropRatio.x4 = Size(41.25,55)
class DropColor extends StatelessWidget {
  const DropColor({
    Key key,
    this.colors,
    this.size,
    this.axis = Axis.vertical
  }) :
   assert(colors != null, "Color must be not-null, DropColor"),
   assert(colors.length == 4, "Colors length must be equal to 4"),
   super(key: key);

  final List<MaterialColor> colors;
  final DropRatio size;
  final Axis  axis;

  static const double defaultRatioHeight = 50.0;
  static const double defaultRatioWidth = 37.5;
  @override
  Widget build(BuildContext context) {
    return Stack(
      alignment: Alignment.center,
      children: [
          for(var index = colors.length-1; index >= 0; index--)...[
            Positioned(
              left: axis == Axis.horizontal ? index * (size.value.width - size.value.width/2.5) : null,
              top:axis == Axis.vertical ? index * (size.value.height - size.value.height/2.5) : null,
              child: CustomPaint(
                painter: _DropColorPainter(colors.reversed.toList()[index]),
                size: size.value
              ),
            )
          ]
        ],
    );
  }
}
/// Drop size must be *3/4* aspect ratio, for keeping drop proportions
/// 
/// **x1** = Size(18.75,25), **x2** = Size(37.5,50)
/// 
/// **x3** = Size(75,100), **x4** = Size(150,200), 
///
/// **x5** = Size(300,400), *default* : x2
enum DropRatio{
  x1,
  x2,
  x3,
  x4,
  x5
}
double calc(x){return x * (3/4);}
/// return **Size** of *enum DropRatio*
/// by default return **DropRatio.x4**
/// 
/// double calc(x){return x * (3/4)} where x is Size(calc(x), x) keeping 3/4 ratio
/// 
/// ```
///   x1 -> x = 25 -> Size(calc(25),25); 
///   x2 -> x = 35 -> Size(calc(35),35);
///   x3 -> x = 45 -> Size(calc(45),45);
///   x4 -> x = 55 -> Size(calc(55),55);
///   x5 -> x = 65 -> Size(calc(65),65);
/// ```
extension DropRatioSize on DropRatio{
  Size get value {
    switch(this){
      case DropRatio.x1:
        return Size(calc(25),25);
        // return Size(18.75,25);
        break;
      case DropRatio.x2:
        return Size(calc(35),35);
        // return Size(26.25,35);
        break;
      case DropRatio.x3:
        return Size(calc(45),45);
        // return Size(33.75,45);
        break;
      case DropRatio.x4:
        return Size(calc(55),55);
        // return Size(41.25,55);
        break;
      case DropRatio.x5:
        return Size(calc(65),65);
        // return Size(48.75,65);
        break;
      default:
        return DropRatio.x4.value;
        break;
    }
  }
}

class _DropColorPainter extends CustomPainter{

  final Color color;

  _DropColorPainter(this.color):assert(color != null,"Color must be not-null, DropColor");
  
  @override
  void paint(Canvas canvas, [Size size = const Size(150,200)]) {
    
    Paint paint = new Paint()
      ..color = this.color
      ..style = PaintingStyle.fill
      ..strokeWidth = 1;
        
    Path path = Path();
    path.moveTo(size.width*0.50,0);
    path.quadraticBezierTo(size.width*-0.00,size.height*0.37,0,size.height*0.63);
    path.cubicTo(size.width*-0.00,size.height*0.85,size.width*0.27,size.height*1.00,size.width*0.50,size.height);
    path.cubicTo(size.width*0.74,size.height*1.00,size.width*1.00,size.height*0.85,size.width,size.height*0.63);
    path.quadraticBezierTo(size.width*1.00,size.height*0.38,size.width*0.50,0);
    path.close();

    
    canvas.drawPath(path, paint);
  }

  @override
  bool shouldRepaint(covariant CustomPainter oldDelegate) {
    return true;
  }
  
}


project/example/lib/pages/all.lol.colors.page.dart
import 'package:flutter/material.dart';
import 'package:lol_colors_flutter/lol_colors_flutter.dart';

import '../drop.widget.dart';
import 'color.variation.page.dart';

final bgColor = Colors.white;
class AllLolColorPage extends StatelessWidget {

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: bgColor,
      appBar: PreferredSize(
        preferredSize: Size.fromHeight(0),
        child: AppBar(
          backgroundColor: bgColor,
          elevation: 0,
        ),
      ),
      body: Column(
        crossAxisAlignment: CrossAxisAlignment.center,
        children: [
          Container(
            height: 40,
            child: Center(child: Text("All Lol Colors", style: TextStyle(fontSize: 20),)),
          ),
          Expanded(
            child: GridView.count(
              crossAxisCount: 3,
              crossAxisSpacing: 0,
              childAspectRatio: 3/4,
              mainAxisSpacing: 5,

              children:all_lol_colors.keys.toList().map((key) 
                => Stack(
                  alignment: Alignment.center,
                  children: [
                    GestureDetector(
                      onTap: ()=>goToColorVariantions(context,key),
                      child: DropColor(
                          colors: all_lol_colors[key],
                          size: DropRatio.x4,
                          // axis: Axis.horizontal,
                          ),
                    ),
                    Positioned(
                      bottom: 0,
                      child: Text(key)
                    )
                  ],
                ),)
                .toList()

                // scrollDirection: Axis.vertical,
                // gridDelegate: SliverGridDelegateWithFixedCrossAxisCount(
                //   crossAxisCount: 5,
                //   mainAxisSpacing: 10
                // ),
              ),
          ),
        ],
      )
   );
  }

  void goToColorVariantions(BuildContext context, String key){
    Navigator.of(context).push(
      MaterialPageRoute(builder: (_) 
        => ColorVariationPage(colors: all_lol_colors[key], numberColor: key,)
      )
    );
  }
}


project/example/lib/pages/color.variation.page.dart
import 'dart:math';

import 'package:flutter/material.dart';
import 'package:lol_colors_flutter/color_extension.dart';

class ColorVariationPage extends StatefulWidget {
  final List<MaterialColor> colors;
  final String numberColor;

  const ColorVariationPage({Key key, this.colors, this.numberColor}) :
   assert(colors != null),
   assert(numberColor != null && numberColor != ''),
   super(key: key);

  @override
  _ColorVariationPageState createState() => _ColorVariationPageState();
}

class _ColorVariationPageState extends State<ColorVariationPage> {
  @override
  Widget build(BuildContext context) {
    final pickRandomColor = widget.colors[range(0,widget.colors.length)];
    final bool isDark = pickRandomColor.isDark();
    final Color dark = Color(0xff131313);
    final Color light = Colors.white;
    return Scaffold(
      appBar: AppBar(
        backgroundColor: pickRandomColor,
        elevation: 0,
        brightness: isDark ? Brightness.light : Brightness.dark,
        title: Text("Color ${widget.numberColor}", style: TextStyle(
          color: isDark ? dark : light
        ),),
        iconTheme: IconThemeData(
          color: isDark ? dark : light
        ),
        actions: [
          IconButton(
            icon: Icon(Icons.refresh),
            onPressed: (){
              setState((){});
            }
          )
        ],
      ),
      body: SingleChildScrollView(
        child: Padding(
          padding: const EdgeInsets.symmetric(vertical: 20, horizontal: 5),
          child: Column(
            children: widget.colors.reversed.map((color) => Column(
              children: [
                BoxColor(color: color,),
                Text(color.toString()),
              ],
            )).toList(),
          ),
        ),
      )
   );
  }
}

final shades = [900,800,700,600,500,400,300,200,100,50];

class BoxColor extends StatelessWidget {
  final MaterialColor color;

  const BoxColor({Key key,@required this.color}) : super(key: key);
  @override
  Widget build(BuildContext context) {
    return Container(
      height: 50,
      width: double.infinity,
      child: ListView.builder(
        scrollDirection: Axis.horizontal,
        itemCount: shades.length,
        itemBuilder: (context, index) {
        return Column(
          mainAxisSize: MainAxisSize.min,
          children: [
            Container(
              color: color[shades[index]],
              height: 40,
              width: 40,
            ),
            color.value.compareTo(color[shades[index]].value) == 0 ?
              Text("${shades[index]}", style: TextStyle(fontSize: 8, fontWeight: FontWeight.w900),):
              Text("${shades[index]}", style: TextStyle(fontSize: 8,),),
          ],
        );
      }),
    );
  }
}

final Random random = Random();
int range(int min, int max){
  return random.nextInt(max) + min;
}


project/example/lib/main.dart
import 'package:example/pages/all.lol.colors.page.dart';
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';
import 'package:lol_colors_flutter/lol_colors_flutter.dart';
import 'package:lol_colors_flutter/color_extension.dart';
import 'drop.widget.dart';

void main() {
  SystemChrome.setSystemUIOverlayStyle(
    SystemUiOverlayStyle(
      statusBarColor: Colors.transparent
    )
  );
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: LolColors.c4714_1,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: HomePage(),
    );
  }
}

class HomePage extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    // print(LolColors.c1278_1[700].computeLuminance());
    // print(values(LolColors.c1278_1[700]));
    return Scaffold(
      appBar: AppBar(
        elevation: 0,
        title: Text("Package Name", style: TextStyle(
          color: Theme.of(context).primaryColor.isDark() ? Colors.black : Colors.white
        ),),
        // backgroundColor: LolColors.color_4714_1[600],
      ),
      body: Column(
        children: [
          Container(
            color: LolColors.c1070_3,
          ),
          Container(
            height: 150,
            width: 150,
            child: Center(child: Text("😂", style: TextStyle(color: Colors.green,fontSize: 50),)),
            decoration: BoxDecoration(
              color: LolColors.c1070_3,
            ),
          ),
          Container(color: LolColors.c4714_1,),
          // Expanded(child: Container(),),
          Container(
            height: 250,
            child: ListView(
              scrollDirection: Axis.horizontal,
                children: all_lol_colors.keys.toList().map((e) 
                => Container(
                  // height: 250,
                  width: 50,
                  child: DropColor(colors: all_lol_colors[e], size: DropRatio.x2, )),)
                .toList(),
            ),
          ),
          TextButton.icon(
            onPressed: (){
              Navigator.of(context).push(MaterialPageRoute(
                builder: (_)=>AllLolColorPage()
              ));
            },
            label: Text("all colors"),
            icon: Icon(Icons.collections),
            style: TextButton.styleFrom(
              primary: LolColors.c1294_3
            ),
          ),
          // Expanded(
          //   child: ListView(
          //     scrollDirection: Axis.vertical,
          //     children: all_lol_colors.keys.toList().map((e) 
          //       => DropColor(colors: all_lol_colors[e],size: Size(18.75,25), ),)
          //       .toList()
          //   ),
          // ),
          // DropColor(colors: all_lol_colors["4714"],size: Size(18.75,25), ),
          Container(
            width: 250,
            height: 50,
            color: LolColors.c2877_3[700],
            child: Center(
              child: Text("Text inside", style: TextStyle(
                color: LolColors.c2877_3[700].isDark() ? Colors.black : Colors.white
              ),),
            ),
          ),
          Container(
            width: 250,
            height: 50,
            color: LolColors.c2877_3[700].negate(),
            child: Center(
              child: Text("Text inside", style: TextStyle(
                color: LolColors.c2877_3[700].negate().isDark() ? Colors.black : Colors.white
              ),),
            ),
          ),
          Container(
            width: 250,
            height: 50,
            color: Colors.black.negate(),
            child: Center(
              child: Text("Text inside", style: TextStyle(
                color: Colors.black.negate().isDark() ? Colors.black : Colors.white
              ),),
            ),
          ),
        ],
      ),
    );
  }
}

final List<Color> colors = [
  Color(0xffffffff),
  Color(0xfffff33f),
  Color(0xff000cc0),
  Color(0xffcceefe),
];

String values(Color color){
  // color.opacity;
  return 
  '''
    toString()    : ${color.toString()},
    color.value   : ${color.value},
    color.opacity : ${color.opacity},
    toRBG         : ${color.toRBG()}
   ''';
}
